File: utils/nfcService.ts
// FurInventory Pro - NFC Service Utility
import NfcManager, { Ndef, NfcTech } from 'react-native-nfc-manager';

class NfcService {
  constructor() {
    this.init();
  }

  async init() {
    try {
      await NfcManager.start();
    } catch (ex) {
      console.warn('NFC Manager initialization failed', ex);
    }
  }

  async isSupported() {
    return await NfcManager.isSupported();
  }

  async isEnabled() {
    return await NfcManager.isEnabled();
  }

  cancel() {
    NfcManager.cancelTechnologyRequest();
  }

  /**
   * Passo 1: Pulisci il Tag
   * Scrive un record NDEF vuoto per "resettare" il tag NFC.
   */
  async cleanTag(): Promise<boolean> {
    let result = false;
    try {
      // Request Ndef technology (works on both iOS and Android)
      await NfcManager.requestTechnology(NfcTech.Ndef, {
        alertMessage: 'Avvicina il tag NFC per pulirlo (1/2)'
      });

      // Write an empty NDEF message to clear the tag
      // TNF_EMPTY is 0
      const emptyRec = Ndef.record(0, '', '', '');
      const bytes = Ndef.encodeMessage([emptyRec]);
      
      if (bytes) {
        await NfcManager.ndefHandler.writeNdefMessage(bytes);
        result = true;
        
        // Optional: show a success message on iOS before closing
        if (NfcManager.setAlertMessageIOS) {
          await NfcManager.setAlertMessageIOS('Tag pulito con successo!');
        }
      }
    } catch (ex) {
      console.warn('Error cleaning tag', ex);
    } finally {
      NfcManager.cancelTechnologyRequest();
    }
    return result;
  }

  /**
   * Passo 2: Scrivi il Tag
   * Scrive un record NDEF URI per registrare il Digital Link.
   */
  async writeGS1Uri(uri: string): Promise<boolean> {
    let result = false;
    try {
      await NfcManager.requestTechnology(NfcTech.Ndef, {
        alertMessage: 'Avvicina il tag NFC per registrare il prodotto (2/2)'
      });

      const bytes = Ndef.encodeMessage([
        Ndef.uriRecord(uri),
      ]);

      if (bytes) {
        await NfcManager.ndefHandler.writeNdefMessage(bytes);
        result = true;

        if (NfcManager.setAlertMessageIOS) {
          await NfcManager.setAlertMessageIOS('Prodotto registrato sul tag!');
        }
      }
    } catch (ex) {
      console.warn('Error writing tag', ex);
    } finally {
      NfcManager.cancelTechnologyRequest();
    }
    return result;
  }

  /**
   * Legge un tag NFC in primo piano e restituisce la stringa contenuta (se URI o Testo)
   */
  async readTag(): Promise<string | null> {
    let tagValue: string | null = null;
    try {
      await NfcManager.requestTechnology(NfcTech.Ndef, {
        alertMessage: 'Avvicina il tag NFC da scansionare'
      });

      const tag = await NfcManager.getTag();
      
      if (tag && tag.ndefMessage && tag.ndefMessage.length > 0) {
        const ndefRecord = tag.ndefMessage[0];
        const payload = new Uint8Array(ndefRecord.payload as number[]);
        // Decode URI
        tagValue = Ndef.uri.decodePayload(payload);
        
        // Se non è un URI, prova a decodificarlo come Testo
        if (!tagValue) {
            tagValue = Ndef.text.decodePayload(payload);
        }
      }
      
      if (NfcManager.setAlertMessageIOS && tagValue) {
        await NfcManager.setAlertMessageIOS('Tag letto con successo!');
      }
    } catch (ex) {
      console.warn('Error reading tag', ex);
    } finally {
      NfcManager.cancelTechnologyRequest();
    }
    
    return tagValue;
  }
}

export const nfcService = new NfcService();

File: flutter_app/lib/utils/nfc_service.dart
import 'dart:typed_data';
import 'package:nfc_manager/nfc_manager.dart';

class NfcService {
  static Future<bool> isSupported() async {
    try {
      return await NfcManager.instance.isAvailable();
    } catch (_) {
      return false;
    }
  }

  static Future<void> start() async {
    try {
      await NfcManager.instance.startSession();
    } catch (_) {}
  }

  static Future<void> stop() async {
    try {
      await NfcManager.instance.stopSession();
    } catch (_) {}
  }

  /// Read an NFC tag and return decoded URI or text.
  static Future<String?> readTag({String alertMessage = 'Avvicina il tag NFC da scansionare'}) async {
    String? tagValue;
    try {
      await NfcManager.instance.startSession(
        alertMessage: alertMessage,
        onDiscovered: (NfcTag tag) async {
          final ndef = Ndef.from(tag);
          if (ndef == null || ndef.cachedMessage == null) {
            await NfcManager.instance.stopSession(errorMessage: 'Tag non NDEF');
            return;
          }
          final records = ndef.cachedMessage!.records;
          if (records.isEmpty) {
            await NfcManager.instance.stopSession(errorMessage: 'Tag vuoto');
            return;
          }
          final record = records.first;
          final payload = record.payload;
          tagValue = _decodePayload(payload, record.typeNameFormat);
          await NfcManager.instance.stopSession(alertMessage: 'Tag letto con successo!');
        },
      );
    } catch (_) {
      try { await NfcManager.instance.stopSession(); } catch (_) {}
    }
    return tagValue;
  }

  /// Clean (reset) an NFC tag by writing an empty NDEF record.
  static Future<bool> cleanTag({String alertMessage = 'Avvicina il tag NFC per pulirlo'}) async {
    var result = false;
    try {
      await NfcManager.instance.startSession(
        alertMessage: alertMessage,
        onDiscovered: (NfcTag tag) async {
          final ndef = Ndef.from(tag);
          if (ndef == null) {
            await NfcManager.instance.stopSession(errorMessage: 'Tag non NDEF');
            return;
          }
          final emptyRecord = NdefRecord(
            typeNameFormat: NdefTypeNameFormat.empty,
            type: Uint8List(0),
            identifier: Uint8List(0),
            payload: Uint8List(0),
          );
          final message = NdefMessage([emptyRecord]);
          await ndef.write(message);
          result = true;
          await NfcManager.instance.stopSession(alertMessage: 'Tag pulito con successo!');
        },
      );
    } catch (_) {
      try { await NfcManager.instance.stopSession(); } catch (_) {}
    }
    return result;
  }

  /// Write a GS1 Digital Link URI to an NFC tag.
  static Future<bool> writeGS1Uri(String uri, {String alertMessage = 'Avvicina il tag NFC per registrare il prodotto'}) async {
    var result = false;
    try {
      await NfcManager.instance.startSession(
        alertMessage: alertMessage,
        onDiscovered: (NfcTag tag) async {
          final ndef = Ndef.from(tag);
          if (ndef == null) {
            await NfcManager.instance.stopSession(errorMessage: 'Tag non NDEF');
            return;
          }
          final uriRecord = NdefRecord.createUri(uri);
          final message = NdefMessage([uriRecord]);
          await ndef.write(message);
          result = true;
          await NfcManager.instance.stopSession(alertMessage: 'Prodotto registrato sul tag!');
        },
      );
    } catch (_) {
      try { await NfcManager.instance.stopSession(); } catch (_) {}
    }
    return result;
  }

  static String? _decodePayload(Uint8List payload, NdefTypeNameFormat tnf) {
    try {
      if (tnf == NdefTypeNameFormat.nfcWellKnown) {
        // URI record: first byte is URI prefix code
        if (payload.isNotEmpty) {
          final prefixCode = payload[0];
          const prefixes = [
            '', 'http://www.', 'https://www.', 'http://', 'https://',
            'tel:', 'mailto:', 'ftp://', 'ftp://', 'ftp://', 'smb://',
            'nfs://', 'ftp://', 'dav://', 'news:', 'telnet://', 'imap:',
            'rtsp://', 'urn:', 'pop:', 'sip:', 'sips://', 'tftp://',
          ];
          final prefix = prefixCode < prefixes.length ? prefixes[prefixCode] : '';
          final rest = String.fromCharCodes(payload.sublist(1));
          return '$prefix$rest';
        }
      } else if (tnf == NdefTypeNameFormat.media || tnf == NdefTypeNameFormat.absoluteUri) {
        return String.fromCharCodes(payload);
      } else {
        return String.fromCharCodes(payload);
      }
    } catch (_) {}
    return null;
  }
}
